### DISCUSSION TITLE
How to access residual vector when doing computeJacobian()

### DISCUSSION THREAD

**Discussion by doquang** [idaholab/moose#26896] [Q&A General] [answered]
Hi,

I am trying to extract the computed residual vector from the scope of class BC::computeJacobian(), but I am getting an error message when using this command `_sys.getVector(_sys.residualVectorTag());`
Can someone help?

Thanks in advance,
M.

```
libMesh terminating:
Cannot retreive vector with tag 1 in system 'nl0'
because a vector has not been associated with that tag.
 ```

**Comment by GiudGiud** [comment]
@YaqiWang and @NamjaeChoi just looked at tags, do you know?

**Reply by NamjaeChoi** [reply]
It seems that the tag gets associated with the residual vector at `FEProblemBase::computeResidual` and `FEProblemBase::computeResidualAndJacobian`. But not sure if your call to `BC::computeJacobian` went through these functions.

**Reply by lindsayad** [reply]
Right, the `residualVectorTag()` is not persistently associated with the residual vector because we could in theory be processing a different residual vector for different invocations of `computeResidualXXX`

**Reply by lindsayad** [reply]
@NamjaeChoi referred to code like this:
```c++
void
FEProblemBase::computeResidual(const NumericVector<Number> & soln,
                               NumericVector<Number> & residual,
                               const unsigned int nl_sys_num)
{
  setCurrentNonlinearSystem(nl_sys_num);

  // We associate the residual tag with the given residual vector to make sure we
  // don't filter it out below
  _current_nl_sys->associateVectorToTag(residual, _current_nl_sys->residualVectorTag());
```
which indeed is not called during Jacobian evaluation

**Reply by doquang** [reply]
Can I get / set the residual values from `NonlinearSystemBase::RHS()` ?

**Reply by lindsayad** [reply, marked as answer]
why do you need this? Setting the residual values during Jacobian evaluation is something like undefined behavior. 

`RHS()` will give you the residual vector, yes

**Reply by doquang** [reply]
I am trying to impose a strong free-slip BC, and I would like to estimate the larger shear stress on the two tangent directions. That why I need to modify both Jacobian & residual.

**Reply by GiudGiud** [reply]
if you modify the residual during the residual loop and the Jacobian during the Jacobian loop you should not have to use the tags.

**Reply by doquang** [reply]
Thanks @GiudGiud. It is what I am doing for now. It is too difficult, and it may not correct for changing both jacobian and residual in the same time, when using SNES solver.

**Reply by GiudGiud** [reply]
ok feel free to use RHS in the residual loop
the tag should also work in the residual loop
